File: cogs/tickets.py
import discord
from discord.ext import commands
from discord import app_commands, ButtonStyle, ui
import re
from db.ticket_db import TicketDB  # your existing DB helper


class Tickets(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.db = TicketDB()

    async def send_log(self, guild, message):
        s = self.db.get_guild_settings(guild.id)
        if not s:
            return
        log_channel = guild.get_channel(s['log_channel'])
        if log_channel:
            await log_channel.send(message)

    async def create_ticket_from_modal(self, interaction: discord.Interaction, description: str):
        settings = self.db.get_guild_settings(interaction.guild.id)
        if not settings:
            return
        category = interaction.guild.get_channel(settings['category_open'])
        if not category:
            return

        overwrites = {
            interaction.guild.default_role: discord.PermissionOverwrite(view_channel=False),
            interaction.user: discord.PermissionOverwrite(view_channel=True, send_messages=True, attach_files=True, embed_links=True),
        }
        for rid in settings['support_roles']:
            role = interaction.guild.get_role(rid)
            if role:
                overwrites[role] = discord.PermissionOverwrite(view_channel=True, send_messages=True)

        channel = await interaction.guild.create_text_channel(
            f"ticket-{interaction.user.name}".lower(),
            overwrites=overwrites,
            category=category,
            reason=f"Ticket opened by {interaction.user}"
        )

        self.db.create_ticket(channel.id, interaction.guild.id, interaction.user.id, [interaction.user.id])

        embed = discord.Embed(
            title="New Ticket",
            description=description,
            color=0x2ecc71
        )
        embed.set_author(name=str(interaction.user), icon_url=interaction.user.display_avatar.url)
        embed.set_footer(text="Staff will respond shortly.")

        await channel.send(content=interaction.user.mention, embed=embed, view=TicketControlView(self))

        await self.send_log(interaction.guild, f"🎟 Ticket {channel.mention} created by {interaction.user.mention}")

    @app_commands.command(name="setticket")
    @app_commands.default_permissions(administrator=True)
    async def setticket(self, interaction: discord.Interaction,
                        ticket_category: discord.CategoryChannel,
                        archive_category: discord.CategoryChannel,
                        support_roles: str,
                        log_channel: discord.TextChannel):
        """Configure ticket system categories, roles, logs"""
        role_ids = []
        for rid in support_roles.split(","):
            rid = rid.strip()
            match = re.search(r"\d+", rid)
            if match:
                role_ids.append(int(match.group()))
        self.db.set_guild_settings(
            interaction.guild.id,
            ticket_category.id,
            archive_category.id,
            role_ids,
            log_channel.id
        )
        await interaction.response.send_message("Ticket system configured.", ephemeral=True)

    @app_commands.command(name="createpanel")
    @app_commands.default_permissions(administrator=True)
    async def createpanel(self, interaction: discord.Interaction,
                          channel: discord.TextChannel,
                          title: str,
                          description: str):
        """Sends the ticket creation panel"""
        embed = discord.Embed(title=title, description=description, color=0x3498db)
        view = TicketPanelView(self.bot)
        await channel.send(embed=embed, view=view)
        await interaction.response.send_message(f"Panel sent to {channel.mention}", ephemeral=True)


class TicketPanelView(discord.ui.View):
    def __init__(self, bot):
        super().__init__(timeout=None)
        self.bot = bot

    @discord.ui.button(label="Open Ticket", style=ButtonStyle.green, custom_id="open_ticket_button")
    async def open_ticket(self, interaction: discord.Interaction, button: discord.ui.Button):
        await interaction.response.send_modal(TicketIssueModal(self.bot))


class TicketIssueModal(discord.ui.Modal, title="Open Support Ticket"):
    def __init__(self, bot):
        super().__init__()
        self.bot = bot

        self.description = ui.TextInput(
            label="Describe your issue/request",
            style=discord.TextStyle.paragraph,
            required=True,
            max_length=1000,
            placeholder="Please provide as much detail as possible so staff can help you."
        )
        self.add_item(self.description)

    async def on_submit(self, interaction: discord.Interaction):
        cog = self.bot.get_cog("Tickets")
        await cog.create_ticket_from_modal(interaction, self.description.value)
        await interaction.response.send_message("Your ticket has been created!", ephemeral=True)


class TicketControlView(discord.ui.View):
    def __init__(self, cog):
        super().__init__(timeout=None)
        self.cog = cog

    @discord.ui.button(label="Claim", style=ButtonStyle.primary, custom_id="ticket_claim")
    async def claim(self, interaction: discord.Interaction, button):
        ticket = self.cog.db.get_ticket(interaction.channel.id)
        if not ticket:
            await interaction.response.send_message("Not a ticket channel.", ephemeral=True)
            return
        support_ids = self.cog.db.get_guild_settings(interaction.guild.id)['support_roles']
        if not any(r.id in support_ids for r in interaction.user.roles):
            await interaction.response.send_message("You are not support staff.", ephemeral=True)
            return
        self.cog.db.update_ticket(interaction.channel.id, claimed_by=interaction.user.id)
        await interaction.response.send_message(f"Ticket claimed by {interaction.user.mention}.", ephemeral=True)
        await self.cog.send_log(interaction.guild, f"📝 Ticket {interaction.channel.mention} claimed by {interaction.user.mention}")

    @discord.ui.button(label="Close", style=ButtonStyle.red, custom_id="ticket_close")
    async def close(self, interaction: discord.Interaction, button):
        ticket = self.cog.db.get_ticket(interaction.channel.id)
        if not ticket:
            await interaction.response.send_message("Not a ticket channel.", ephemeral=True)
            return
        settings = self.cog.db.get_guild_settings(interaction.guild.id)
        archive_category = interaction.guild.get_channel(settings['category_archive'])
        if not archive_category:
            await interaction.response.send_message("Archive category missing.", ephemeral=True)
            return
        await interaction.channel.edit(category=archive_category, reason="Ticket closed")
        self.cog.db.update_ticket(interaction.channel.id, status='closed')
        await interaction.response.send_message("Ticket closed and archived.", ephemeral=True)
        await self.cog.send_log(interaction.guild, f"📁 Ticket {interaction.channel.name} closed by {interaction.user.mention}")

    @discord.ui.button(label="Delete", style=ButtonStyle.danger, custom_id="ticket_delete")
    async def delete(self, interaction: discord.Interaction, button):
        ticket = self.cog.db.get_ticket(interaction.channel.id)
        if not ticket:
            await interaction.response.send_message("Not a ticket channel.", ephemeral=True)
            return
        support_ids = self.cog.db.get_guild_settings(interaction.guild.id)['support_roles']
        if interaction.user.id != ticket['claimed_by'] and not any(r.id in support_ids for r in interaction.user.roles):
            await interaction.response.send_message("You must claim or be support to delete.", ephemeral=True)
            return
        await self.cog.send_log(interaction.guild, f"❌ Ticket {interaction.channel.name} deleted by {interaction.user.mention}")
        self.cog.db.delete_ticket(interaction.channel.id)
        await interaction.channel.delete(reason=f"Deleted by {interaction.user}")


async def setup(bot):
    cog = Tickets(bot)
    await bot.add_cog(cog)

    # Register persistent views ONCE, for post-reboot working buttons
    bot.add_view(TicketPanelView(bot))
    bot.add_view(TicketControlView(cog))

File: commands/tickets.js
const { 
    SlashCommandBuilder,
    PermissionFlagsBits,
    ChannelType,
    EmbedBuilder
} = require('discord.js');
const ticketModel = require('../models/ticketModel');
const { TicketPanelView, TicketControlView } = require('../utils/ticketComponents');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('tickets')
        .setDescription('Manage the ticket system')
        .setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild)
        .addSubcommand(subcommand =>
            subcommand
                .setName('setup')
                .setDescription('Configure the ticket system')
                .addChannelOption(option =>
                    option.setName('open_category')
                        .setDescription('Category for open tickets')
                        .setRequired(true)
                        .addChannelTypes(ChannelType.GuildCategory))
                .addChannelOption(option =>
                    option.setName('archive_category')
                        .setDescription('Category for closed tickets')
                        .setRequired(true)
                        .addChannelTypes(ChannelType.GuildCategory))
                .addRoleOption(option =>
                    option.setName('support_role')
                        .setDescription('Role for support staff who can manage tickets')
                        .setRequired(true))
                .addChannelOption(option =>
                    option.setName('log_channel')
                        .setDescription('Channel for ticket logs')
                        .setRequired(true)
                        .addChannelTypes(ChannelType.GuildText)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('panel')
                .setDescription('Create a ticket panel with an open button')
                .addChannelOption(option =>
                    option.setName('channel')
                        .setDescription('Channel to post the ticket panel in')
                        .setRequired(true)
                        .addChannelTypes(ChannelType.GuildText))
                .addStringOption(option =>
                    option.setName('title')
                        .setDescription('Title for the ticket panel')
                        .setRequired(true))
                .addStringOption(option =>
                    option.setName('description')
                        .setDescription('Description for the ticket panel')
                        .setRequired(true)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('add')
                .setDescription('Add a user to the current ticket')
                .addUserOption(option =>
                    option.setName('user')
                        .setDescription('User to add to the ticket')
                        .setRequired(true)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('remove')
                .setDescription('Remove a user from the current ticket')
                .addUserOption(option =>
                    option.setName('user')
                        .setDescription('User to remove from the ticket')
                        .setRequired(true))),

    async execute(interaction) {
        const subcommand = interaction.options.getSubcommand();

        switch (subcommand) {
            case 'setup':
                await this.setupTickets(interaction);
                break;
            case 'panel':
                await this.createPanel(interaction);
                break;
            case 'add':
                await this.addUserToTicket(interaction);
                break;
            case 'remove':
                await this.removeUserFromTicket(interaction);
                break;
            default:
                await interaction.reply({
                    content: 'Unknown subcommand.',
                    flags: 64
                });
        }
    },

    async setupTickets(interaction) {
        const openCategory = interaction.options.getChannel('open_category');
        const archiveCategory = interaction.options.getChannel('archive_category');
        const supportRole = interaction.options.getRole('support_role');
        const logChannel = interaction.options.getChannel('log_channel');

        // Save settings to database
        await ticketModel.setGuildSettings(
            interaction.guild.id,
            openCategory.id,
            archiveCategory.id,
            [supportRole.id],
            logChannel.id
        );

        await interaction.reply({
            content: `✅ Ticket system configured:\n- Open tickets category: ${openCategory}\n- Archive category: ${archiveCategory}\n- Support role: ${supportRole}\n- Log channel: ${logChannel}`,
            flags: 64
        });
    },

    async createPanel(interaction) {
        const channel = interaction.options.getChannel('channel');
        const title = interaction.options.getString('title');
        const description = interaction.options.getString('description');

        // Create embed for ticket panel
        const embed = new EmbedBuilder()
            .setTitle(title)
            .setDescription(description)
            .setColor(0x2B82CB)
            .setFooter({ 
                text: `${interaction.guild.name} • Support Tickets`, 
                iconURL: interaction.guild.iconURL() 
            })
            .setTimestamp();

        // Create panel with ticket button
        const ticketPanel = new TicketPanelView();

        // Send the panel to the specified channel
        await channel.send({
            embeds: [embed],
            components: ticketPanel.components
        });

        await interaction.reply({
            content: `✅ Ticket panel created in ${channel}`,
            flags: 64
        });
    },

    async createTicketFromModal(client, interaction, description) {
        // Get guild settings
        const settings = await ticketModel.getGuildSettings(interaction.guildId);
        if (!settings) {
            return await interaction.reply({
                content: 'Ticket system is not configured. Please contact an administrator.',
                flags: 64
            });
        }

        // Get the open tickets category
        const category = interaction.guild.channels.cache.get(settings.categoryOpen);
        if (!category) {
            return await interaction.reply({
                content: 'Tickets category not found. Please contact an administrator.',
                flags: 64
            });
        }

        // Check if user already has an open ticket
        const userTickets = await ticketModel.getUserTickets(interaction.guildId, interaction.user.id);
        const openUserTickets = userTickets.filter(t => t.status === 'open');
        
        if (openUserTickets.length > 0) {
            const existingTicket = interaction.guild.channels.cache.get(openUserTickets[0].channelId);
            if (existingTicket) {
                return await interaction.reply({
                    content: `You already have an open ticket: ${existingTicket}`,
                    flags: 64
                });
            }
        }

        // Create a unique ticket name
        const ticketNumber = Math.floor(Math.random() * 10000);
        const ticketName = `ticket-${interaction.user.username}-${ticketNumber}`;

        try {
            // Create ticket channel
            const ticketChannel = await interaction.guild.channels.create({
                name: ticketName,
                type: ChannelType.GuildText,
                parent: category.id,
                permissionOverwrites: [
                    {
                        id: interaction.guild.id,
                        deny: [PermissionFlagsBits.ViewChannel]
                    },
                    {
                        id: interaction.user.id,
                        allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages]
                    }
                ]
            });

            // Add permissions for support roles
            for (const roleId of settings.supportRoles) {
                await ticketChannel.permissionOverwrites.create(roleId, {
                    ViewChannel: true,
                    SendMessages: true,
                    ReadMessageHistory: true
                });
            }

            // Create ticket in database
            await ticketModel.createTicket(
                ticketChannel.id,
                interaction.guildId,
                interaction.user.id,
                [interaction.user.id]
            );

            // Create welcome message
            const welcomeEmbed = new EmbedBuilder()
                .setTitle('New Support Ticket')
                .setDescription(`
                    Thank you for creating a ticket ${interaction.user}!
                    
                    **Issue Description:**
                    ${description}
                    
                    A support team member will assist you shortly.
                    Please provide any additional information that might help us.
                `)
                .setColor(0x2B82CB)
                .setTimestamp();

            // Add ticket control buttons
            const ticketControls = new TicketControlView();

            // Send welcome message with buttons
            await ticketChannel.send({
                content: `${interaction.user}`,
                embeds: [welcomeEmbed],
                components: ticketControls.components
            });

            // Reply to the original interaction
            await interaction.reply({
                content: `✅ Your ticket has been created: ${ticketChannel}`,
                flags: 64
            });

            // Log ticket creation with enhanced embed
            await this.sendLog(
                client,
                interaction.guildId,
                `🎫 New ticket created by ${interaction.user} in ${ticketChannel}`,
                {
                    color: 0x2ECC71, // Green
                    author: {
                        name: `Ticket Created | ${interaction.user.tag}`,
                        iconURL: interaction.user.displayAvatarURL()
                    },
                    fields: [
                        { name: 'Ticket', value: ticketChannel.name, inline: true },
                        { name: 'Created By', value: `${interaction.user}`, inline: true },
                        { name: 'Description', value: description.substring(0, 100) + (description.length > 100 ? '...' : ''), inline: false }
                    ]
                }
            );
        } catch (error) {
            console.error('Error creating ticket:', error);
            await interaction.reply({
                content: 'Failed to create ticket. Please try again later.',
                flags: 64
            });
        }
    },

    async addUserToTicket(interaction) {
        // Get ticket data
        const ticket = await ticketModel.getTicket(interaction.channelId);
        if (!ticket) {
            return await interaction.reply({
                content: 'This command can only be used in a ticket channel.',
                flags: 64
            });
        }

        const user = interaction.options.getUser('user');

        // Check if user is already in the ticket
        if (ticket.participants.includes(user.id)) {
            return await interaction.reply({
                content: `${user} is already in this ticket.`,
                flags: 64
            });
        }

        // Add user to channel permissions
        await interaction.channel.permissionOverwrites.create(user.id, {
            ViewChannel: true,
            SendMessages: true,
            ReadMessageHistory: true
        });

        // Update participants in database
        ticket.participants.push(user.id);
        await ticketModel.updateTicket(interaction.channelId, {
            participants: ticket.participants
        });

        await interaction.reply({
            content: `✅ Added ${user} to the ticket.`,
            flags: 0
        });
    },

    async removeUserFromTicket(interaction) {
        // Get ticket data
        const ticket = await ticketModel.getTicket(interaction.channelId);
        if (!ticket) {
            return await interaction.reply({
                content: 'This command can only be used in a ticket channel.',
                flags: 64
            });
        }

        const user = interaction.options.getUser('user');

        // Check if user is the creator (can't remove ticket creator)
        if (user.id === ticket.creatorId) {
            return await interaction.reply({
                content: `You cannot remove the ticket creator.`,
                flags: 64
            });
        }

        // Check if user is in the ticket
        if (!ticket.participants.includes(user.id)) {
            return await interaction.reply({
                content: `${user} is not in this ticket.`,
                flags: 64
            });
        }

        // Remove user from channel permissions
        await interaction.channel.permissionOverwrites.delete(user.id);

        // Update participants in database
        const updatedParticipants = ticket.participants.filter(id => id !== user.id);
        await ticketModel.updateTicket(interaction.channelId, {
            participants: updatedParticipants
        });

        await interaction.reply({
            content: `✅ Removed ${user} from the ticket.`,
            flags: 0
        });
    },

    async sendLog(client, guildId, message, options = {}) {
        try {
            // Get guild settings
            const settings = await ticketModel.getGuildSettings(guildId);
            if (!settings || !settings.logChannel) return;

            // Get log channel
            const logChannel = client.channels.cache.get(settings.logChannel);
            if (!logChannel) return;

            // Create embed
            const embed = new EmbedBuilder()
                .setDescription(message)
                .setColor(options.color || 0x2B82CB)
                .setTimestamp();
            
            // Add author if provided
            if (options.author) {
                embed.setAuthor({ 
                    name: options.author.name || 'Ticket System',
                    iconURL: options.author.iconURL
                });
            } else {
                embed.setAuthor({ 
                    name: 'Ticket System',
                    iconURL: client.user.displayAvatarURL()
                });
            }
            
            // Add footer if provided or default
            if (options.footer) {
                embed.setFooter({ text: options.footer });
            } else {
                embed.setFooter({ text: `Guild ID: ${guildId}` });
            }
            
            // Add fields if provided
            if (options.fields && Array.isArray(options.fields)) {
                embed.addFields(options.fields);
            }

            // Send log message with embed
            await logChannel.send({ embeds: [embed] });
        } catch (error) {
            console.error('Error sending ticket log message:', error);
        }
    }
};